### PR TITLE
DNS Forward Zone Resource and Data Source Implementation

### PR DESCRIPTION
**Pull request by unasra**

### CODE REVIEW

File: internal/service/dns_config/api_forward_zone_data_source.go [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+	resp.Schema = schema.Schema{
+		MarkdownDescription: ``,
+		Attributes: map[string]schema.Attribute{
+			"filters": schema.MapAttribute{
+				Description: "Filter are used to return a more specific list of results. Filters can be used to match resources by specific attributes, e.g. name. If you specify multiple filters, the results returned will have only resources that match all the specified filters.",
+				ElementType: types.StringType,
+				Optional:    true,
+			},
+			"tag_filters": schema.MapAttribute{
+				Description: "Tag Filters are used to filter by tags return a more specific list of results. If you specify multiple filters, the results returned will have only resources that match all the specified filters.",
```
Comment (mathewab) [inline internal/service/dns_config/api_forward_zone_data_source.go:57]:
```suggestion
				Description: "Tag Filters are used to return a more specific list of results filtered by tags. If you specify multiple filters, the results returned will have only resources that match all the specified filters.",
```

File: internal/service/dns_config/api_forward_zone_data_source_test.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+	resourceName := "bloxone_dns_forward_zone.test"
+	var v dns_config.ConfigForwardZone
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:                 func() { acctest.PreCheck(t) },
+		ProtoV6ProviderFactories: acctest.ProtoV6ProviderFactories,
+		CheckDestroy:             testAccCheckForwardZoneDestroy(context.Background(), &v),
+		Steps: []resource.TestStep{
+			{
+				Config: testAccForwardZoneDataSourceConfigFilters("FQDN_REPLACE_ME"),
```
Comment (mathewab) [inline internal/service/dns_config/api_forward_zone_data_source_test.go:25]: can you pls replace REPLACE_ME?  :)

File: internal/service/dns_config/api_forward_zone_resource.go [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+	client *bloxoneclient.APIClient
+}
+
+func (r *ForwardZoneResource) Metadata(ctx context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_" + "dns_forward_zone"
+}
+
+func (r *ForwardZoneResource) Schema(ctx context.Context, req resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		MarkdownDescription: ``,
```
Comment (mathewab) [inline internal/service/dns_config/api_forward_zone_resource.go:34]:
```suggestion
		MarkdownDescription: "This object (dns/forward_zone) represents a forwarding zone",
```

File: internal/service/dns_config/api_forward_zone_resource_test.go [hunk 4]
```diff
@@ -0,0 +1,553 @@
...
+	var resourceName = "bloxone_dns_forward_zone.test"
+	var v dns_config.ConfigForwardZone
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:                 func() { acctest.PreCheck(t) },
+		ProtoV6ProviderFactories: acctest.ProtoV6ProviderFactories,
+		Steps: []resource.TestStep{
+			// Create and Read
+			{
+				Config: testAccForwardZoneBasicConfig("tf-acc-test.com."),
```
Comment (mathewab) [inline internal/service/dns_config/api_forward_zone_resource_test.go:33]: can you use a random name for all these tests?

File: internal/service/dns_config/api_forward_zone_resource_test.go [hunk 5]
```diff
@@ -0,0 +1,553 @@
...
+	resource.Test(t, resource.TestCase{
+		PreCheck:                 func() { acctest.PreCheck(t) },
+		ProtoV6ProviderFactories: acctest.ProtoV6ProviderFactories,
+		Steps: []resource.TestStep{
+			// Create and Read
+			{
+				Config: testAccForwardZoneHosts("tf-acc-test.com."),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckForwardZoneExists(context.Background(), resourceName, &v),
+					resource.TestCheckResourceAttr(resourceName, "fqdn", "tf-acc-test.com."),
```
Comment (mathewab) [inline internal/service/dns_config/api_forward_zone_resource_test.go:231]: shouldn't you be testing hosts here?

File: internal/service/dns_config/api_forward_zone_resource_test.go [hunk 6]
```diff
@@ -0,0 +1,553 @@
...
+	resource.Test(t, resource.TestCase{
+		PreCheck:                 func() { acctest.PreCheck(t) },
+		ProtoV6ProviderFactories: acctest.ProtoV6ProviderFactories,
+		Steps: []resource.TestStep{
+			// Create and Read
+			{
+				Config: testAccForwardZoneInternalForwarders("tf-acc-test.com."),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckForwardZoneExists(context.Background(), resourceName, &v),
+					resource.TestCheckResourceAttr(resourceName, "fqdn", "tf-acc-test.com."),
```
Comment (mathewab) [inline internal/service/dns_config/api_forward_zone_resource_test.go:260]: same comment as above, shouldn't you be testing hosts here?

File: internal/service/dns_config/api_forward_zone_resource_test.go [hunk 7]
```diff
@@ -0,0 +1,553 @@
...
+			Execute()
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+}
+
+func testAccForwardZoneBasicConfig(fqdn string) string {
+	// TODO: create basic resource with required fields
```
Comment (mathewab) [inline internal/service/dns_config/api_forward_zone_resource_test.go:425]: can you please remove TODO

File: internal/service/dns_config/api_forward_zone_resource_test.go [hunk 8]
```diff
@@ -0,0 +1,553 @@
...
+resource "bloxone_dns_forward_zone" "test_forward_only" {
+    fqdn = %q
+    forward_only = %q
+}
+`, fqdn, forwardOnly)
+}
+
+func testAccBaseWithHost() string {
+	return `
+	data "bloxone_dns_hosts" "all_hosts" {
```
Comment (mathewab) [inline internal/service/dns_config/api_forward_zone_resource_test.go:476]:
```suggestion
	data "bloxone_dns_hosts" "test_host_01" {
```

File: internal/service/dns_config/api_forward_zone_resource_test.go [hunk 9]
```diff
@@ -0,0 +1,553 @@
...
+
+resource "bloxone_dns_forward_nsg" "two"{
+	name = %q
+}
+
+resource "bloxone_dns_forward_zone" "test_nsgs" {
+	fqdn = %q
+    nsgs = [%s.id]
+}
+`, acctest.RandomNameWithPrefix("auth-nsg"), acctest.RandomNameWithPrefix("auth-nsg"), fqdn, nsgs)
```
Comment (mathewab) [inline internal/service/dns_config/api_forward_zone_resource_test.go:518]:
```suggestion
`, acctest.RandomNameWithPrefix("fw-nsg"), acctest.RandomNameWithPrefix("fw-nsg"), fqdn, nsgs)
```

File: internal/service/dns_config/api_forward_zone_resource_test.go [hunk 10]
```diff
@@ -0,0 +1,553 @@
...
+	name = %q
+}
+
+resource "bloxone_dns_forward_nsg" "two"{
+	name = %q
+}
+
+resource "bloxone_dns_forward_zone" "test_nsgs" {
+	fqdn = %q
+    nsgs = [%s.id]
```
Comment (mathewab) [inline internal/service/dns_config/api_forward_zone_resource_test.go:516]: formatting, probably a whitespace vs tab issue

File: internal/service/dns_config/api_forward_zone_resource_test.go [hunk 11]
```diff
@@ -0,0 +1,553 @@
...
+	name = %q
+}
+
+resource "bloxone_dns_view" "two" {
+	name = %q
+}
+
+resource "bloxone_dns_forward_zone" "test_view" {
+	fqdn = %q
+    view = %s.id
```
Comment (mathewab) [inline internal/service/dns_config/api_forward_zone_resource_test.go:550]: formatting, probably a whitespace vs tab issue

File: internal/service/dns_config/api_forward_zone_resource_test.go [hunk 12]
```diff
@@ -0,0 +1,553 @@
...
+`, fqdn, forwardOnly)
+}
+
+func testAccBaseWithHost() string {
+	return `
+	data "bloxone_dns_hosts" "all_hosts" {
+		filters = {
+			name = "TF_TEST_HOST_01"
+	}
+	}
```
Comment (mathewab) [inline internal/service/dns_config/api_forward_zone_resource_test.go:480]: formatting, probably a whitespace vs tab issue

File: internal/service/dns_config/model_config_forward_zone.go [hunk 13]
```diff
@@ -0,0 +1,224 @@
...
+	"mapping": schema.StringAttribute{
+		Computed:            true,
+		MarkdownDescription: "Read-only. Zone mapping type. Allowed values:  * _forward_,  * _ipv4_reverse_.  * _ipv6_reverse_.  Defaults to _forward_.",
+	},
+	"nsgs": schema.ListAttribute{
+		ElementType: types.StringType,
+		Optional:    true,
+		PlanModifiers: []planmodifier.List{
+			listplanmodifier.RequiresReplaceIfConfigured(),
+		},
```
Comment (mathewab) [inline internal/service/dns_config/model_config_forward_zone.go:128]: why do you need this for nsgs?

File: internal/service/dns_config/model_config_forward_zone.go [hunk 14]
```diff
@@ -0,0 +1,224 @@
...
+		MarkdownDescription: "The resource identifier.",
+	},
+	"parent": schema.StringAttribute{
+		Optional:            true,
+		MarkdownDescription: "The resource identifier.",
+	},
```
Comment (mathewab) [inline internal/service/dns_config/model_config_forward_zone.go:134]: parent is a read only field

File: internal/service/dns_config/model_config_host.go [hunk 15]
```diff
@@ -2,6 +2,8 @@ package dns_config
 
 import (
 	"context"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/planmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/stringplanmodifier"
 
 	"github.com/hashicorp/terraform-plugin-framework/attr"
 	"github.com/hashicorp/terraform-plugin-framework/diag"
```
Comment (mathewab) [inline internal/service/dns_config/model_config_host.go:9]: imports not sorted

File: internal/service/dns_config/model_config_forward_zone_config.go [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+	schema "github.com/hashicorp/terraform-plugin-framework/resource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/hashicorp/terraform-plugin-framework/types/basetypes"
+
+	"github.com/infobloxopen/bloxone-go-client/dns_config"
+
+	"github.com/infobloxopen/terraform-provider-bloxone/internal/flex"
+)
+
+type ConfigForwardZoneConfigModel struct {
```
Comment (mathewab) [inline internal/service/dns_config/model_config_forward_zone_config.go:17]: i don't think this model is used. you can remove it.

File: internal/service/dns_config/api_forward_zone_data_source.go (hunk 1 above)
Comment (unasra) [reply internal/service/dns_config/api_forward_zone_data_source.go]: Resolved

File: internal/service/dns_config/api_forward_zone_data_source_test.go (hunk 2 above)
Comment (unasra) [reply internal/service/dns_config/api_forward_zone_data_source_test.go]: Resolved

File: internal/service/dns_config/api_forward_zone_resource_test.go (hunk 4 above)
Comment (unasra) [reply internal/service/dns_config/api_forward_zone_resource_test.go]: Resolved

File: internal/service/dns_config/api_forward_zone_resource_test.go (hunk 6 above)
Comment (unasra) [reply internal/service/dns_config/api_forward_zone_resource_test.go]: Resolved

File: internal/service/dns_config/api_forward_zone_resource_test.go (hunk 7 above)
Comment (unasra) [reply internal/service/dns_config/api_forward_zone_resource_test.go]: Resolved

File: internal/service/dns_config/api_forward_zone_resource_test.go (hunk 8 above)
Comment (unasra) [reply internal/service/dns_config/api_forward_zone_resource_test.go]: Resolved

File: internal/service/dns_config/api_forward_zone_resource_test.go (hunk 9 above)
Comment (unasra) [reply internal/service/dns_config/api_forward_zone_resource_test.go]: Resolved

File: internal/service/dns_config/api_forward_zone_resource_test.go (hunk 10 above)
Comment (unasra) [reply internal/service/dns_config/api_forward_zone_resource_test.go]: Resolved

File: internal/service/dns_config/api_forward_zone_resource_test.go (hunk 11 above)
Comment (unasra) [reply internal/service/dns_config/api_forward_zone_resource_test.go]: Resolved

File: internal/service/dns_config/api_forward_zone_resource_test.go (hunk 12 above)
Comment (unasra) [reply internal/service/dns_config/api_forward_zone_resource_test.go]: Resolved

File: internal/service/dns_config/model_config_forward_zone.go (hunk 13 above)
Comment (unasra) [reply internal/service/dns_config/model_config_forward_zone.go]: Resolved

File: internal/service/dns_config/model_config_host.go (hunk 15 above)
Comment (unasra) [reply internal/service/dns_config/model_config_host.go]: Resolved

File: internal/service/dns_config/model_config_forward_zone_config.go (hunk 16 above)
Comment (unasra) [reply internal/service/dns_config/model_config_forward_zone_config.go]: Resolved

File: examples/data-sources/bloxone_dns_forward_zones/data-source.tf [hunk 17]
```diff
@@ -0,0 +1,16 @@
...
+}
+
+# Get forward zones filtered by tag
+data "bloxone_dns_forward_zones" "example_by_tag" {
+  tag_filters = {
+    region = "eu"
+  }
+}
+
+# Get all auth zones
```
Comment (venkat-iblox) [inline examples/data-sources/bloxone_dns_forward_zones/data-source.tf:15]:
```suggestion
# Get all forward zones
```

File: examples/resources/bloxone_dns_forward_zone/resource.tf [hunk 18]
```diff
@@ -0,0 +1,15 @@
+resource "bloxone_dns_view" "example_view" {
+  name = "example_dns_view"
+}
+
+resource "bloxone_dns_forward_zone" "example" {
+  fqdn = "tf-acc-test.com."
+
+  # Other optional fields
+  name    = "example_forward_zone"
+  comment = "Example of an Forward Zone"
```
Comment (venkat-iblox) [inline examples/resources/bloxone_dns_forward_zone/resource.tf:10]:
```suggestion
  comment = "Example of a Forward Zone"
```

File: internal/service/dns_config/api_forward_zone_data_source_test.go [hunk 19]
```diff
@@ -0,0 +1,114 @@
...
+					append([]resource.TestCheckFunc{
+						testAccCheckForwardZoneExists(context.Background(), resourceName, &v),
+					}, testAccCheckForwardZoneResourceAttrPair(resourceName, dataSourceName)...)...,
+				),
+			},
+		},
+	})
+}
+
+// below all TestAcc functions
```
Comment (venkat-iblox) [inline internal/service/dns_config/api_forward_zone_data_source_test.go:60]: Comment can be removed
```suggestion
```

File: internal/service/dns_config/api_view_resource.go [hunk 20]
```diff
@@ -3,11 +3,10 @@ package dns_config
 import (
 	"context"
 	"fmt"
-	"net/http"
-
 	"github.com/hashicorp/terraform-plugin-framework/path"
 	"github.com/hashicorp/terraform-plugin-framework/resource"
 	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
+	"net/http"
```
Comment (venkat-iblox) [inline internal/service/dns_config/api_view_resource.go:9]: Please fix import ordering

File: examples/data-sources/bloxone_dns_forward_zones/data-source.tf (hunk 17 above)
Comment (unasra) [reply examples/data-sources/bloxone_dns_forward_zones/data-source.tf]: Resolved

File: examples/resources/bloxone_dns_forward_zone/resource.tf (hunk 18 above)
Comment (unasra) [reply examples/resources/bloxone_dns_forward_zone/resource.tf]: Resolved

File: internal/service/dns_config/api_forward_zone_data_source_test.go (hunk 19 above)
Comment (unasra) [reply internal/service/dns_config/api_forward_zone_data_source_test.go]: Resolved

File: internal/service/dns_config/api_view_resource.go (hunk 20 above)
Comment (unasra) [reply internal/service/dns_config/api_view_resource.go]: Resolved

File: examples/resources/bloxone_dns_forward_zone/resource.tf [hunk 21]
```diff
@@ -0,0 +1,15 @@
+resource "bloxone_dns_view" "example_view" {
+  name = "example_dns_view"
+}
+
+resource "bloxone_dns_forward_zone" "example" {
+  fqdn = "tf-acc-test.com."
+
+  # Other optional fields
+  name    = "example_forward_zone"
```
Comment (mathewab) [inline examples/resources/bloxone_dns_forward_zone/resource.tf:9]: I don't think there is a name attribute in forward zone?

File: internal/service/dns_config/api_forward_zone_resource_test.go [hunk 22]
```diff
@@ -0,0 +1,568 @@
...
+			},
+			// Update and Read
+			{
+				Config: testAccForwardZoneHosts(fqdn),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckForwardZoneExists(context.Background(), resourceName, &v),
+					resource.TestCheckResourceAttr(resourceName, "hosts.#", "1"),
+					resource.TestCheckResourceAttrPair(resourceName, "hosts.0", "data.bloxone_dns_hosts.test_host_01", "results.0.id"),
+				),
+			},
```
Comment (mathewab) [inline internal/service/dns_config/api_forward_zone_resource_test.go:251]: I don't see any updates here. If there is nothing to be updated than this case can be removed or you can put a to do item for updating the host when you have the second host available.

File: internal/service/dns_config/api_forward_zone_resource_test.go [hunk 23]
```diff
@@ -0,0 +1,568 @@
...
+			// Update and Read
+			{
+				Config: testAccForwardZoneInternalForwarders(fqdn),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckForwardZoneExists(context.Background(), resourceName, &v),
+					resource.TestCheckResourceAttr(resourceName, "internal_forwarders.#", "1"),
+					resource.TestCheckResourceAttrPair(resourceName, "internal_forwarders.0", "data.bloxone_dns_hosts.test_host_01", "results.0.id"),
+				),
+			},
```
Comment (mathewab) [inline internal/service/dns_config/api_forward_zone_resource_test.go:283]: same comment as above 

I don't see any updates here. If there is nothing to be updated than this case can be removed or you can put a to do item for updating the host when you have the second host available.

Comment (unasra) [reply internal/service/dns_config/api_forward_zone_resource_test.go]: fixed

File: internal/service/dns_config/api_forward_zone_resource_test.go (hunk 22 above)
Comment (unasra) [reply internal/service/dns_config/api_forward_zone_resource_test.go]: fixed

File: examples/resources/bloxone_dns_forward_zone/resource.tf (hunk 21 above)
Comment (unasra) [reply examples/resources/bloxone_dns_forward_zone/resource.tf]: yeah no name is there . This field is removed

**Review by mathewab — APPROVED** [review]

**Comment by mathewab** [comment]
Hey, @venkat-iblox sorry I didn't notice that you hadn't responded to the resolution of some of your comments. I should have waited before merging :) @unasra if there's any more comments form venkat please do in another PR. Thanks.

**Comment by unasra** [comment]
@mathewab I have addressed all the conversations by @venkat-iblox . The issues are still unresolved , if he has any changes I will address them in another PR.

**Comment by venkat-iblox** [comment]
> @mathewab I have addressed all the conversations by @venkat-iblox . The issues are still unresolved , if he has any changes I will address them in another PR.

No further comments.

**Pull request merged**